File: bot/src/audio/queue.ts
export enum PlayMode {
  Sequential = "seq",
  Loop = "loop",
  Random = "random",
  RandomLoop = "rloop",
}

/**
 * Who put the track on the queue.
 * - `user` / omitted: human request (!add, web, tools) — jumps ahead of radio fill
 * - `radio`: auto-DJ / profile restock filler
 * - `system`: bot-internal (rarely used)
 */
export type QueueSource = "user" | "radio" | "system";

export interface QueuedSong {
  id: string;
  name: string;
  artist: string;
  album: string;
  platform: "local" | "youtube" | "stream";
  url?: string; // resolved lazily at play time
  coverUrl: string;
  duration: number; // seconds
  /** Defaults to human priority when omitted. */
  source?: QueueSource;
}

/** Auto-DJ fill tracks lose to any human add. */
export function isRadioFill(song: QueuedSong): boolean {
  return song.source === "radio";
}

export class PlayQueue {
  private songs: QueuedSong[] = [];
  private currentIndex = -1;
  private mode: PlayMode = PlayMode.Sequential;
  private playedIndices = new Set<number>();
  private history: number[] = [];
  private forwardStack: number[] = [];
  private static readonly HISTORY_LIMIT = 50;

  private pushHistory(idx: number): void {
    if (idx < 0 || idx >= this.songs.length) return;
    this.history.push(idx);
    if (this.history.length > PlayQueue.HISTORY_LIMIT) {
      this.history.shift();
    }
  }

  /**
   * Enqueue a song.
   * - **Radio fill** (`source: "radio"`): always append at the end.
   * - **Human / default**: insert after the current track and after any other
   *   human-priority tracks, but **before** the first radio-fill track so
   *   `!add` jumps the auto-DJ queue.
   * @returns index where the song was placed
   */
  add(song: QueuedSong): number {
    if (isRadioFill(song) || this.songs.length === 0) {
      this.songs.push(song);
      return this.songs.length - 1;
    }

    // After now-playing (or at 0 if nothing active), skip past human tracks,
    // stop at the first radio filler.
    const start = this.currentIndex < 0 ? 0 : this.currentIndex + 1;
    let insertAt = this.songs.length;
    for (let i = start; i < this.songs.length; i++) {
      if (isRadioFill(this.songs[i]!)) {
        insertAt = i;
        break;
      }
    }

    if (insertAt >= this.songs.length) {
      this.songs.push(song);
      return this.songs.length - 1;
    }

    this.songs.splice(insertAt, 0, song);
    this.shiftIndicesAfterInsert(insertAt);
    return insertAt;
  }

  addMany(songs: QueuedSong[]): void {
    for (const s of songs) this.add(s);
  }

  /**
   * Insert a song to play immediately after the current one. Falls
   * through to plain push when nothing is playing yet (currentIndex < 0
   * or queue empty), so the existing "add → idle bot starts playing"
   * flow continues to work.
   *
   * Shifts playedIndices and history entries > currentIndex by +1 so
   * their references stay valid after the splice.
   */
  addNext(song: QueuedSong): void {
    const withSource: QueuedSong = song.source ? song : { ...song, source: "user" };
    if (this.currentIndex < 0 || this.songs.length === 0) {
      this.songs.push(withSource);
      return;
    }
    const insertAt = this.currentIndex + 1;
    this.songs.splice(insertAt, 0, withSource);
    this.shiftIndicesAfterInsert(insertAt);
  }

  /** After splicing a song at `insertAt`, shift bookkeeping indices ≥ insertAt. */
  private shiftIndicesAfterInsert(insertAt: number): void {
    const shifted = new Set<number>();
    for (const i of this.playedIndices) {
      shifted.add(i >= insertAt ? i + 1 : i);
    }
    this.playedIndices = shifted;
    this.history = this.history.map((i) => (i >= insertAt ? i + 1 : i));
    this.forwardStack = this.forwardStack.map((i) => (i >= insertAt ? i + 1 : i));
    // currentIndex is never ≥ insertAt when we insert after current; keep as-is.
  }

  remove(index: number): QueuedSong | null {
    if (index < 0 || index >= this.songs.length) return null;
    const [removed] = this.songs.splice(index, 1);

    if (index < this.currentIndex) {
      this.currentIndex--;
    } else if (index === this.currentIndex) {
      this.currentIndex--;
    }

    // Rebuild playedIndices to account for shifted indices
    const newPlayed = new Set<number>();
    for (const idx of this.playedIndices) {
      if (idx === index) continue;
      newPlayed.add(idx > index ? idx - 1 : idx);
    }
    this.playedIndices = newPlayed;

    // Same shift logic for history — drop entries pointing at the
    // removed song; shift entries > index down by 1.
    this.history = this.history
      .filter((idx) => idx !== index)
      .map((idx) => (idx > index ? idx - 1 : idx));

    // forwardStack (prev→next resume) must shift the same way or next()
    // can land on a stale index and return undefined.
    this.forwardStack = this.forwardStack
      .filter((idx) => idx !== index)
      .map((idx) => (idx > index ? idx - 1 : idx));

    return removed;
  }

  clear(): void {
    this.songs = [];
    this.currentIndex = -1;
    this.playedIndices.clear();
    this.history = [];
    this.forwardStack = [];
  }

  play(): QueuedSong | null {
    if (this.songs.length === 0) return null;
    this.playedIndices.clear();
    this.history = [];
    this.forwardStack = [];
    this.currentIndex = 0;
    this.playedIndices.add(0);
    return this.songs[0];
  }

  playAt(index: number): QueuedSong | null {
    if (index < 0 || index >= this.songs.length) return null;
    this.pushHistory(this.currentIndex);
    // Reset the Random-mode "unplayed" pool — explicit picks restart
    // shuffle from this point. History tracking is independent and
    // unaffected by this clear.
    this.playedIndices.clear();
    this.forwardStack = [];
    this.currentIndex = index;
    this.playedIndices.add(index);
    return this.songs[index];
  }

  next(): QueuedSong | null {
    if (this.songs.length === 0) return null;

    switch (this.mode) {
      case PlayMode.Sequential: {
        const nextIndex = this.currentIndex + 1;
        if (nextIndex >= this.songs.length) return null;
        this.pushHistory(this.currentIndex);
        this.currentIndex = nextIndex;
        return this.songs[nextIndex];
      }
      case PlayMode.Loop: {
        this.pushHistory(this.currentIndex);
        this.currentIndex = (this.currentIndex + 1) % this.songs.length;
        return this.songs[this.currentIndex];
      }
      case PlayMode.Random:
      case PlayMode.RandomLoop: {
        // Prefer returning to the position recorded in the forward stack (prev song).
        // Skip stale / OOB entries (defensive; remove/addNext also shift the stack).
        while (this.forwardStack.length > 0) {
          const target = this.forwardStack.pop()!;
          if (target < 0 || target >= this.songs.length || target === this.currentIndex) {
            continue;
          }
          this.pushHistory(this.currentIndex);
          this.currentIndex = target;
          this.playedIndices.add(target);
          return this.songs[target];
        }

        // Shuffle bag: pick uniformly from the songs not yet played this
        // cycle, so every song plays once before any repeats.
        // Songs added mid-cycle aren't in playedIndices, so they're
        // naturally eligible within the current cycle.
        const unplayed: number[] = [];
        for (let i = 0; i < this.songs.length; i++) {
          if (!this.playedIndices.has(i)) unplayed.push(i);
        }

        if (unplayed.length === 0) {
          // Cycle complete.
          if (this.mode === PlayMode.Random) return null; // Random: stop after current
          // Random loop: reshuffle and keep going forever.
          if (this.songs.length === 1) {
            this.pushHistory(this.currentIndex);
            this.currentIndex = 0;
            this.playedIndices = new Set([0]);
            return this.songs[0];
          }
          // Start a fresh cycle: every song is eligible again, but exclude
          // the song that just played from THIS pick only, so it doesn't
          // repeat back-to-back across the boundary. It stays eligible for
          // the rest of the new cycle, so every song still plays exactly once.
          this.playedIndices = new Set();
          for (let i = 0; i < this.songs.length; i++) {
            if (i !== this.currentIndex) unplayed.push(i);
          }
        }

        const nextIndex = unplayed[Math.floor(Math.random() * unplayed.length)];
        this.pushHistory(this.currentIndex);
        this.currentIndex = nextIndex;
        this.playedIndices.add(nextIndex);
        return this.songs[nextIndex];
      }
    }
  }

  prev(): QueuedSong | null {
    if (this.songs.length === 0) return null;

    // Record current position to forward stack for next to prefer returning
    if (this.currentIndex >= 0 && this.forwardStack.length < PlayQueue.HISTORY_LIMIT) {
      this.forwardStack.push(this.currentIndex);
    }

    // Preferred: pop from the back-stack so prev means "the song I
    // actually played before this one," not "the previous array slot."
    while (this.history.length > 0) {
      const idx = this.history.pop()!;
      if (idx >= 0 && idx < this.songs.length) {
        this.currentIndex = idx;
        this.playedIndices = new Set([...this.history, this.currentIndex]);
        return this.songs[idx];
      }
      // Stale entry (song removed) — keep popping.
    }

    // Fallback: no history to walk back through. In Sequential we
    // can still meaningfully step the index backward; in random
    // modes there's nothing useful to return.
    if (this.mode === PlayMode.Random || this.mode === PlayMode.RandomLoop) {
      return null;
    }
    const prevIndex = this.currentIndex - 1;
    if (prevIndex < 0) {
      if (this.mode === PlayMode.Sequential) return null;
      this.currentIndex = this.songs.length - 1;
    } else {
      this.currentIndex = prevIndex;
    }
    this.playedIndices.add(this.currentIndex);
    return this.songs[this.currentIndex];
  }

  current(): QueuedSong | null {
    if (this.currentIndex < 0 || this.currentIndex >= this.songs.length) return null;
    return this.songs[this.currentIndex];
  }

  list(): QueuedSong[] {
    return [...this.songs];
  }

  size(): number {
    return this.songs.length;
  }

  isEmpty(): boolean {
    return this.songs.length === 0;
  }

  getMode(): PlayMode {
    return this.mode;
  }

  setMode(mode: PlayMode): void {
    this.mode = mode;
    this.playedIndices.clear();
    this.history = [];
    this.forwardStack = [];
    if (this.currentIndex >= 0) {
      this.playedIndices.add(this.currentIndex);
    }
  }

  getCurrentIndex(): number {
    return this.currentIndex;
  }

  /** Number of songs not yet played in Random mode. */
  unplayedCount(): number {
    return this.songs.length - this.playedIndices.size;
  }
}

File: crates/mp-music/src/queue.rs
// Copyright (c) 2026 Lane Ambrose
// SPDX-License-Identifier: MIT

//! PlayQueue — port of `bot/src/audio/queue.ts`.
//! Default mode is RandomLoop. Commands that REPLACE the queue must call
//! [`PlayQueue::set_mode`] with an explicit mode (usually Sequential) before
//! [`PlayQueue::play`], or a one-song queue loops forever.

use crate::track::QueuedSong;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PlayMode {
    Sequential,
    Loop,
    Random,
    RandomLoop,
}

impl PlayMode {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Sequential => "seq",
            Self::Loop => "loop",
            Self::Random => "random",
            Self::RandomLoop => "rloop",
        }
    }

    pub fn parse(s: &str) -> Option<Self> {
        match s {
            "seq" => Some(Self::Sequential),
            "loop" => Some(Self::Loop),
            "random" => Some(Self::Random),
            "rloop" => Some(Self::RandomLoop),
            _ => None,
        }
    }
}

impl std::fmt::Display for PlayMode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.as_str())
    }
}

pub struct PlayQueue {
    songs: Vec<QueuedSong>,
    current_index: i32,
    mode: PlayMode,
    played_indices: std::collections::HashSet<usize>,
    history: Vec<usize>,
    forward_stack: Vec<usize>,
}

const HISTORY_LIMIT: usize = 50;

impl Default for PlayQueue {
    fn default() -> Self {
        Self::new()
    }
}

impl PlayQueue {
    pub fn new() -> Self {
        Self {
            songs: Vec::new(),
            current_index: -1,
            mode: PlayMode::RandomLoop,
            played_indices: std::collections::HashSet::new(),
            history: Vec::new(),
            forward_stack: Vec::new(),
        }
    }

    fn push_history(&mut self, idx: i32) {
        if idx < 0 {
            return;
        }
        let idx = idx as usize;
        if idx >= self.songs.len() {
            return;
        }
        self.history.push(idx);
        if self.history.len() > HISTORY_LIMIT {
            self.history.remove(0);
        }
    }

    /// Radio fill always appends. Human/default inserts after current and
    /// after other human tracks, before the first radio filler.
    pub fn add(&mut self, song: QueuedSong) -> usize {
        if song.is_radio_fill() || self.songs.is_empty() {
            self.songs.push(song);
            return self.songs.len() - 1;
        }
        let start = if self.current_index < 0 {
            0
        } else {
            self.current_index as usize + 1
        };
        let mut insert_at = self.songs.len();
        for i in start..self.songs.len() {
            if self.songs[i].is_radio_fill() {
                insert_at = i;
                break;
            }
        }
        if insert_at >= self.songs.len() {
            self.songs.push(song);
            return self.songs.len() - 1;
        }
        self.songs.insert(insert_at, song);
        self.shift_indices_after_insert(insert_at);
        insert_at
    }

    pub fn add_many(&mut self, songs: Vec<QueuedSong>) -> i32 {
        let mut first = -1i32;
        for s in songs {
            let at = self.add(s);
            if first < 0 {
                first = at as i32;
            }
        }
        first
    }

    pub fn add_next(&mut self, song: QueuedSong) {
        if self.current_index < 0 || self.songs.is_empty() {
            self.songs.push(song);
            return;
        }
        let insert_at = self.current_index as usize + 1;
        self.songs.insert(insert_at, song);
        self.shift_indices_after_insert(insert_at);
    }

    fn shift_indices_after_insert(&mut self, insert_at: usize) {
        self.played_indices = self
            .played_indices
            .iter()
            .map(|&i| if i >= insert_at { i + 1 } else { i })
            .collect();
        for i in &mut self.history {
            if *i >= insert_at {
                *i += 1;
            }
        }
        for i in &mut self.forward_stack {
            if *i >= insert_at {
                *i += 1;
            }
        }
    }

    pub fn remove(&mut self, index: usize) -> Option<QueuedSong> {
        if index >= self.songs.len() {
            return None;
        }
        let removed = self.songs.remove(index);
        if (index as i32) <= self.current_index {
            self.current_index -= 1;
        }
        self.played_indices = self
            .played_indices
            .iter()
            .filter(|&&i| i != index)
            .map(|&i| if i > index { i - 1 } else { i })
            .collect();
        self.history = self
            .history
            .iter()
            .copied()
            .filter(|&i| i != index)
            .map(|i| if i > index { i - 1 } else { i })
            .collect();
        self.forward_stack = self
            .forward_stack
            .iter()
            .copied()
            .filter(|&i| i != index)
            .map(|i| if i > index { i - 1 } else { i })
            .collect();
        Some(removed)
    }

    pub fn clear(&mut self) {
        self.songs.clear();
        self.current_index = -1;
        self.played_indices.clear();
        self.history.clear();
        self.forward_stack.clear();
    }

    pub fn play(&mut self) -> Option<QueuedSong> {
        if self.songs.is_empty() {
            return None;
        }
        self.played_indices.clear();
        self.history.clear();
        self.forward_stack.clear();
        self.current_index = 0;
        self.played_indices.insert(0);
        Some(self.songs[0].clone())
    }

    pub fn play_at(&mut self, index: usize) -> Option<QueuedSong> {
        if index >= self.songs.len() {
            return None;
        }
        self.push_history(self.current_index);
        self.played_indices.clear();
        self.forward_stack.clear();
        self.current_index = index as i32;
        self.played_indices.insert(index);
        Some(self.songs[index].clone())
    }

    pub fn next(&mut self) -> Option<QueuedSong> {
        if self.songs.is_empty() {
            return None;
        }
        match self.mode {
            PlayMode::Sequential => {
                let next_index = self.current_index + 1;
                if next_index >= self.songs.len() as i32 {
                    return None;
                }
                self.push_history(self.current_index);
                self.current_index = next_index;
                Some(self.songs[next_index as usize].clone())
            }
            PlayMode::Loop => {
                self.push_history(self.current_index);
                self.current_index = (self.current_index + 1) % self.songs.len() as i32;
                Some(self.songs[self.current_index as usize].clone())
            }
            PlayMode::Random | PlayMode::RandomLoop => self.next_random(),
        }
    }

    fn next_random(&mut self) -> Option<QueuedSong> {
        while let Some(target) = self.forward_stack.pop() {
            if target >= self.songs.len() || target as i32 == self.current_index {
                continue;
            }
            self.push_history(self.current_index);
            self.current_index = target as i32;
            self.played_indices.insert(target);
            return Some(self.songs[target].clone());
        }

        let mut unplayed: Vec<usize> = (0..self.songs.len())
            .filter(|i| !self.played_indices.contains(i))
            .collect();

        if unplayed.is_empty() {
            if self.mode == PlayMode::Random {
                return None;
            }
            if self.songs.len() == 1 {
                self.push_history(self.current_index);
                self.current_index = 0;
                self.played_indices = std::collections::HashSet::from([0]);
                return Some(self.songs[0].clone());
            }
            self.played_indices.clear();
            for i in 0..self.songs.len() {
                if i as i32 != self.current_index {
                    unplayed.push(i);
                }
            }
        }

        let pick = rand::random::<usize>() % unplayed.len();
        let next_index = unplayed[pick];
        self.push_history(self.current_index);
        self.current_index = next_index as i32;
        self.played_indices.insert(next_index);
        Some(self.songs[next_index].clone())
    }

    pub fn prev(&mut self) -> Option<QueuedSong> {
        if self.songs.is_empty() {
            return None;
        }
        if self.current_index >= 0 && self.forward_stack.len() < HISTORY_LIMIT {
            self.forward_stack.push(self.current_index as usize);
        }
        while let Some(idx) = self.history.pop() {
            if idx < self.songs.len() {
                self.current_index = idx as i32;
                self.played_indices.insert(idx);
                return Some(self.songs[idx].clone());
            }
        }
        if self.mode == PlayMode::Random || self.mode == PlayMode::RandomLoop {
            return None;
        }
        let prev_index = self.current_index - 1;
        if prev_index < 0 {
            if self.mode == PlayMode::Sequential {
                return None;
            }
            self.current_index = self.songs.len() as i32 - 1;
        } else {
            self.current_index = prev_index;
        }
        self.played_indices.insert(self.current_index as usize);
        Some(self.songs[self.current_index as usize].clone())
    }

    pub fn current(&self) -> Option<QueuedSong> {
        if self.current_index < 0 || self.current_index >= self.songs.len() as i32 {
            return None;
        }
        Some(self.songs[self.current_index as usize].clone())
    }

    pub fn list(&self) -> Vec<QueuedSong> {
        self.songs.clone()
    }

    pub fn size(&self) -> usize {
        self.songs.len()
    }

    pub fn is_empty(&self) -> bool {
        self.songs.is_empty()
    }

    pub fn get_mode(&self) -> PlayMode {
        self.mode
    }

    pub fn set_mode(&mut self, mode: PlayMode) {
        self.mode = mode;
        self.played_indices.clear();
        self.history.clear();
        self.forward_stack.clear();
        if self.current_index >= 0 {
            self.played_indices.insert(self.current_index as usize);
        }
    }

    pub fn get_current_index(&self) -> i32 {
        self.current_index
    }

    pub fn unplayed_count(&self) -> usize {
        self.songs.len().saturating_sub(self.played_indices.len())
    }
}

/// Replace the queue with one track and force Sequential.
/// Any command that clears+refills+plays must go through here (or call
/// `set_mode(Sequential)` itself) so RandomLoop cannot inherit onto a
/// one-song queue.
pub fn replace_queue_with_song(queue: &mut PlayQueue, song: QueuedSong) -> Option<QueuedSong> {
    queue.clear();
    queue.add(song);
    queue.set_mode(PlayMode::Sequential);
    queue.play()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::track::{Platform, QueueSource};

    fn song(id: &str) -> QueuedSong {
        song_named(id, id)
    }
    fn song_named(id: &str, name: &str) -> QueuedSong {
        QueuedSong {
            id: id.into(),
            name: name.into(),
            artist: "Artist".into(),
            album: "Album".into(),
            platform: Platform::Youtube,
            url: format!("https://example.com/{id}.mp3"),
            cover_url: String::new(),
            duration: 240,
            source: QueueSource::User,
        }
    }
    fn radio(id: &str) -> QueuedSong {
        let mut s = song(id);
        s.source = QueueSource::Radio;
        s
    }

    #[test]
    fn starts_empty() {
        let q = PlayQueue::new();
        assert!(q.is_empty());
        assert!(q.current().is_none());
        assert_eq!(q.size(), 0);
    }

    #[test]
    fn user_add_jumps_radio() {
        let mut q = PlayQueue::new();
        q.add(radio("now"));
        q.play();
        q.add(radio("r1"));
        q.add(radio("r2"));
        let at = q.add(song("user1"));
        assert_eq!(at, 1);
        assert_eq!(
            q.list().iter().map(|s| s.id.as_str()).collect::<Vec<_>>(),
            ["now", "user1", "r1", "r2"]
        );
        q.add(song("user2"));
        assert_eq!(
            q.list().iter().map(|s| s.id.as_str()).collect::<Vec<_>>(),
            ["now", "user1", "user2", "r1", "r2"]
        );
        q.add(radio("r3"));
        assert_eq!(
            q.list().iter().map(|s| s.id.as_str()).collect::<Vec<_>>(),
            ["now", "user1", "user2", "r1", "r2", "r3"]
        );
    }

    #[test]
    fn sequential_advances_and_ends() {
        let mut q = PlayQueue::new();
        q.set_mode(PlayMode::Sequential);
        q.add(song("1"));
        q.add(song("2"));
        q.add(song("3"));
        q.play();
        assert_eq!(q.current().unwrap().id, "1");
        assert_eq!(q.next().unwrap().id, "2");
        assert_eq!(q.next().unwrap().id, "3");
        assert!(q.next().is_none());
    }

    #[test]
    fn loop_wraps() {
        let mut q = PlayQueue::new();
        q.set_mode(PlayMode::Loop);
        q.add(song("1"));
        q.play();
        assert_eq!(q.next().unwrap().id, "1");
    }

    #[test]
    fn defaults_to_random_loop() {
        assert_eq!(PlayQueue::new().get_mode(), PlayMode::RandomLoop);
    }

    #[test]
    fn random_loop_one_song_replays() {
        let mut q = PlayQueue::new();
        q.add(song("only"));
        q.play();
        assert_eq!(q.next().unwrap().id, "only");
    }

    #[test]
    fn replace_forces_sequential() {
        let mut q = PlayQueue::new();
        assert_eq!(q.get_mode(), PlayMode::RandomLoop);
        replace_queue_with_song(&mut q, song("a"));
        assert_eq!(q.get_mode(), PlayMode::Sequential);
        assert_eq!(q.current().unwrap().id, "a");
        assert!(q.next().is_none(), "one-song Sequential must run dry");
    }

    #[test]
    fn remove_before_current_shifts() {
        let mut q = PlayQueue::new();
        q.set_mode(PlayMode::Sequential);
        q.add(song("A"));
        q.add(song("B"));
        q.add(song("C"));
        q.play_at(2);
        q.remove(0);
        assert_eq!(q.current().unwrap().id, "C");
        assert_eq!(q.get_current_index(), 1);
    }

    #[test]
    fn remove_current_lets_next_advance() {
        let mut q = PlayQueue::new();
        q.set_mode(PlayMode::Sequential);
        q.add(song("A"));
        q.add(song("B"));
        q.add(song("C"));
        q.add(song("D"));
        q.play_at(2);
        q.remove(2);
        assert_eq!(q.next().unwrap().id, "D");
    }

    #[test]
    fn prev_sequential() {
        let mut q = PlayQueue::new();
        q.set_mode(PlayMode::Sequential);
        q.add(song("1"));
        q.add(song("2"));
        q.play();
        q.next();
        assert_eq!(q.current().unwrap().id, "2");
        q.prev();
        assert_eq!(q.current().unwrap().id, "1");
    }
}
